### PR TITLE
Rundeck integration

### PR DESCRIPTION
**Pull request by gal-berger**
## Status
- [] In Progress
- [X] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/25109

## Minimum version of Demisto
- [ ] 5.0.0
- [X] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 1]
```diff
@@ -0,0 +1,1113 @@
...
+        outputs=filtered_results,
+        outputs_key_field='id'
+    )
+
+
+def adhoc_run_command(client: Client, args: dict):
+    project_name: str = args.get('project_name', '')
+    exec_command: str = args.get('exec_command', '')
+    node_thread_count: str = args.get('node_thread_count', '')
+    node_keepgoing: str = args.get('node_keepgoing', '')  # TODO: add list option true\false
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:888]: done?

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 2]
```diff
@@ -0,0 +1,1113 @@
...
+    as_user: str = args.get('as_user')
+    node_filter: str = args.get('node_filter', '')
+
+    demisto.info('sending adhoc run request')
+    result = client.adhoc_run(project_name, exec_command, node_thread_count, node_keepgoing, as_user, node_filter)
+    demisto.info('finish sending adhoc run request')
+
+    # if not isinstance(result, dict):
+        # raise DemistoException(f'Got unexpected response: {result}')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:897]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 3]
```diff
@@ -0,0 +1,1113 @@
...
+        outputs=filtered_results,
+        outputs_key_field='id'
+    )
+
+
+def adhoc_script_run_command(client: Client, args: dict):
+    project_name: str = args.get('project_name', '')
+    arg_string: str = args.get('arg_string', '')
+    node_thread_count: str = args.get('node_thread_count', '')
+    node_keepgoing: str = args.get('node_keepgoing', '')  # TODO: add list option true\false
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:916]: todo done?

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 4]
```diff
@@ -0,0 +1,1113 @@
...
+    as_user: str = args.get('as_user')
+    script_interpreter: str = args.get('script_interpreter', '')
+    interpreter_args_quoted: str = args.get('interpreter_args_quoted', '')  # TODO: add list option true\false
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:919]: todo done?

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 5]
```diff
@@ -0,0 +1,1113 @@
...
+    file_extension: str = args.get('file_extension', '')
+    node_filter: str = args.get('node_filter', '')
+    entry_id: str = args.get('entry_id', '')
+    demisto.info('sending adhoc script run request')
+    result = client.adhoc_script_run(project_name, arg_string, node_thread_count, node_keepgoing, as_user, node_filter
+                                     , script_interpreter, interpreter_args_quoted, file_extension, entry_id)
+    demisto.info('finish sending adhoc script run request')
+
+    # if not isinstance(result, dict):
+    # raise DemistoException(f'Got unexpected response: {result}')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:929]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 6]
```diff
@@ -0,0 +1,1113 @@
...
+                files={'scriptFile': file},
+                url_suffix=f'/project/{project_name_to_pass}/run/script',
+                params=request_params
+            )
+
+
+''' HELPER FUNCTIONS '''
+
+
+def filter_results(results: Union[list, dict], fields_to_remove: list, remove_signs) -> Union[list, dict]:
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:465]: please remove logs from this function. this is an internal object manipulation function, no need for logging.
as long as its not errors. then we will want to add the input arguments in the error.

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 7]
```diff
@@ -0,0 +1,1113 @@
...
+                    for sign in remove_signs:
+                        if sign in key:
+                            new_record[key.replace(sign, '')] = value
+                        else:
+                            new_record[key] = value
+            new_results.append(new_record)
+    return new_results
+
+
+def attribute_pairs_to_dict(attrs_str: Optional[str], delim_char: str = ","):
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:506]: please remove logs from this function. this is an internal object manipulation function, no need for logging.
as long as its not errors. then we will want to add the input arguments in the error.

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 8]
```diff
@@ -0,0 +1,1113 @@
...
+            raise ValueError(f"Could not parse field: {f}")
+        demisto.debug(f'add this key: {match.group(1)} and this value: {match.group(2)} to attrs')
+        attrs.update({match.group(1): match.group(2)})
+        demisto.debug(f'finish adding this key: {match.group(1)} and this value: {match.group(2)} to attrs')
+
+    return attrs
+
+
+def convert_str_to_int(val_to_convert: Optional[str], param_name: str) -> Optional[int]:
+    demisto.info(f'start converting {val_to_convert} to integer')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:535]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 9]
```diff
@@ -0,0 +1,1113 @@
...
+    if val_to_convert:
+        try:
+            return int(val_to_convert)
+        except ValueError:
+            raise DemistoException(f'\'{param_name}\' most be a number.')
+        except Exception:
+            demisto.error(f'failed to convert {val_to_convert} to integer')
+            raise
+    demisto.info(f'finish converting {val_to_convert} to integer')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:544]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 10]
```diff
@@ -0,0 +1,1113 @@
...
+    :param selected_time: the delta you want to get from today:
+    '1 hour': for one hour from now
+    '1 week': for one week from now
+    '1 day': for one day from now
+    :return: the selected time in ISO-8601 format.
+    """
+    selected_iso_time = ''
+    if not selected_time:
+        return selected_iso_time
+    iso_with_timezone = parse(f'in {selected_time} UTC').isoformat()
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:562]: is this calculating the local time correctly?

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 11]
```diff
@@ -0,0 +1,1113 @@
...
+                                        exclude_group_path_exact, job_filter, exclude_job_filter,
+                                        job_exact_filter, exclude_job_exact_filter, execution_type_filter,
+                                        max_results, offset, project_name)
+    demisto.info('finish sending job execution query request')
+
+    if not isinstance(result, dict):
+        raise DemistoException(f'got unexpected results from api: {result}')
+
+    executions: list = result.get('executions', [])
+    demisto.info('start filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:788]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 12]
```diff
@@ -0,0 +1,1113 @@
...
+    filtered_executions = filter_results(executions, ['href', 'permalink'], ['-'])
+    demisto.info('finish filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:790]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 13]
```diff
@@ -0,0 +1,1113 @@
...
+    execution_id: Optional[int] = convert_str_to_int(args.get('execution_id'), 'execution_id')
+
+    demisto.info('sending job execution abort request')
+    result = client.job_execution_abort(execution_id)
+    demisto.info('finish sending job execution abort request')
+
+    if not isinstance(result, dict):
+        raise DemistoException(f'Got unexpected response: {result}')
+
+    demisto.info('start filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:870]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 14]
```diff
@@ -0,0 +1,1113 @@
...
+    filtered_results = filter_results(result, ['href', 'permalink'], ['-'])
+    demisto.info('finish filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:872]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 15]
```diff
@@ -0,0 +1,1113 @@
...
+    as_user: str = args.get('as_user')
+    node_filter: str = args.get('node_filter', '')
+
+    demisto.info('sending adhoc run request')
+    result = client.adhoc_run(project_name, exec_command, node_thread_count, node_keepgoing, as_user, node_filter)
+    demisto.info('finish sending adhoc run request')
+
+    # if not isinstance(result, dict):
+        # raise DemistoException(f'Got unexpected response: {result}')
+    demisto.info('start filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:898]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 16]
```diff
@@ -0,0 +1,1113 @@
...
+    node_filter: str = args.get('node_filter', '')
+    entry_id: str = args.get('entry_id', '')
+    demisto.info('sending adhoc script run request')
+    result = client.adhoc_script_run(project_name, arg_string, node_thread_count, node_keepgoing, as_user, node_filter
+                                     , script_interpreter, interpreter_args_quoted, file_extension, entry_id)
+    demisto.info('finish sending adhoc script run request')
+
+    # if not isinstance(result, dict):
+    # raise DemistoException(f'Got unexpected response: {result}')
+    demisto.info('start filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:930]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 17]
```diff
@@ -0,0 +1,1113 @@
...
+    filtered_results = filter_results(result, ['href', 'permalink'], ['-'])
+    demisto.info('finish filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:932]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 18]
```diff
@@ -0,0 +1,1113 @@
...
+    node_filter: str = args.get('node_filter', '')
+    arg_string: str = args.get('arg_string', '')
+    demisto.log('sending adhoc script run from url request')
+    result = client.adhoc_script_run_from_url(project_name, script_url, node_thread_count, node_keepgoing, as_user,
+                                              node_filter, script_interpreter, interpreter_args_quoted, file_extension,
+                                              arg_string)
+    demisto.log('finish sending adhoc script run from url request')
+    # if not isinstance(result, dict):
+    # raise DemistoException(f'Got unexpected response: {result}')
+    demisto.log('start filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:963]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 19]
```diff
@@ -0,0 +1,1113 @@
...
+    filtered_results = filter_results(result, ['href', 'permalink'], ['-'])
+    demisto.log('finish filter results from the api')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:965]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 20]
```diff
@@ -0,0 +1,1113 @@
...
+    script_interpreter: str = args.get('script_interpreter', '')
+    interpreter_args_quoted: str = args.get('interpreter_args_quoted', '')
+    file_extension: str = args.get('file_extension', '')
+    node_filter: str = args.get('node_filter', '')
+    arg_string: str = args.get('arg_string', '')
+    demisto.log('sending adhoc script run from url request')
+    result = client.adhoc_script_run_from_url(project_name, script_url, node_thread_count, node_keepgoing, as_user,
+                                              node_filter, script_interpreter, interpreter_args_quoted, file_extension,
+                                              arg_string)
+    demisto.log('finish sending adhoc script run from url request')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:960]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 21]
```diff
@@ -0,0 +1,1113 @@
...
+    try:
+        demisto.info('start convert "options" argument to str')
+        if options_as_dict:
+            options_as_str = json.dumps(options_as_dict)
+        else:
+            options_as_str = options_as_dict
+            demisto.info('finish convert "options" argument to str')
+    except Exception as e:
+        raise DemistoException(f'There was a problem converting "free_json" to json. The reason is: {e}')
+    demisto.info('sending webhook event send request')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:992]: remove

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 22]
```diff
@@ -0,0 +1,1113 @@
...
+    result = client.webhook_event_send(auth_token, options_as_str, free_json)
+    demisto.info('finish sending webhook event send request')
```
Comment (yuvalbenshalom) [inline Packs/Rundeck/Integrations/Rundeck/Rundeck.py:994]: remove

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
mostly cleanup. see my comments.

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py (hunk 1 above)
Comment (gal-berger) [reply Packs/Rundeck/Integrations/Rundeck/Rundeck.py:888]: yes

File: Packs/Rundeck/Integrations/Rundeck/Rundeck.py [hunk 23]
```diff
@@ -0,0 +1,1113 @@
...
+    file_extension: str = args.get('file_extension', '')
+    node_filter: str = args.get('node_filter', '')
+    arg_string: str = args.get('arg_string', '')
+    demisto.log('sending adhoc script run from url request')
+    result = client.adhoc_script_run_from_url(project_name, script_url, node_thread_count, node_keepgoing, as_user,
+                                              node_filter, script_interpreter, interpreter_args_quoted, file_extension,
+                                              arg_string)
+    demisto.log('finish sending adhoc script run from url request')
+    # if not isinstance(result, dict):
+    # raise DemistoException(f'Got unexpected response: {result}')
```
[the remark this answers is not in the document]
Comment (gal-berger) [reply Packs/Rundeck/Integrations/Rundeck/Rundeck.py:962]: I've commented it out

File: Packs/Rundeck/Integrations/Rundeck/Rundeck_description.md [hunk 24]
```diff
@@ -0,0 +1,3 @@
+## Hello World
```
[the remark this answers is not in the document]
Comment (gal-berger) [reply Packs/Rundeck/Integrations/Rundeck/Rundeck_description.md]: you've commented on a file that is not up to date.
please view this file again.

File: Packs/Rundeck/.secrets-ignore [hunk 25]
```diff
@@ -0,0 +1,224 @@
+https://www.markdownguide.org
```
[the remark this answers is not in the document]
Comment (gal-berger) [reply Packs/Rundeck/.secrets-ignore:1]: when you create a new pack using demisto-sdk, it uses it in the code examples.

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
looks good.
only one delete from the yml(in comment)
and add command examples to the readme.

File: Packs/Rundeck/Integrations/Rundeck/README.md [hunk 26]
```diff
@@ -0,0 +1,518 @@
+Rundeck is runbook automation for incident management, business continuity, and self-service operations.
```
[the remark this answers is not in the document]
Comment (gal-berger) [reply Packs/Rundeck/Integrations/Rundeck/README.md:1]: I've put them in "commands_examples.txt" should I add them as well here and delete the examples file?

**Review by yuvalbenshalom — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@yuvalbenshalom Doc review completed.

**Comment by gal-berger** [comment]
@ShirleyDenkberg can you please review this file as well?: `Rundeck-job-execute-Generic_README.md`

**Comment by ShirleyDenkberg** [comment]
@yuvalbenshalom Looks good.

**Pull request merged**